src/api/cvc4cpp.h

```diff
@@ -0,0 +1,2357 @@
+     * @return true if the sort is a bit-vector sort
+     */
+    bool isBitVector() const;
+
+    /**
+     * Is this a floating-point sort?
+     * @return true if the sort is a floating-point sort
+     */
+    bool isFloatingPoint() const;
+
+    /**
+     * Is this a datatype sort?
+     * @return true if the sort is a datatype sort
+     */
+    bool isDatatype() const;
+
+    /**
+     * Is this a function sort?
+     * @return true if the sort is a function sort
+     */
+    bool isFunction() const;
+
+    /**
+     * Is this a predicate sort?
+     * That is, is this a function sort mapping to Boolean? All predicate
+     * sorts are also function sorts.
+     * @return true if the sort is a predicate sort
+     */
+    bool isPredicate() const;
+
+    /**
+     * Is this a tuple sort?
+     * @return true if the sort is a tuple sort
+     */
+    bool isTuple() const;
+
+    /**
+     * Is this a record sort?
+     * @return true if the sort is a record sort
+     */
+    bool isRecord() const;
+
+    /**
+     * Is this an array sort?
+     * @return true if the sort is a array sort
+     */
+    bool isArray() const;
+
+    /**
+     * Is this a Set sort?
+     * @return true if the sort is a Set sort
+     */
+    bool isSet() const;
+
+    /**
+     * Is this a sort kind?
+     * @return true if this is a sort kind
+     */
+    bool isSort() const;
+
+    /**
+     * Is this a sort constructor kind?
+     * @return true if this is a sort constructor kind
+     */
+    bool isSortConstructor() const;
+
+    /**
+     * @return the underlying datatype of a datatype sort
+     */
+    Datatype getDatatype() const;
+
+    /**
+     * Instantiate a parameterized datatype/sort sort.
+     * Create sorts parameter with Solver::mkParamSort().
+     * @param params the list of sort parameters to instantiate with
+     */
+    Sort instantiate(const std::vector<Sort>& params) const;
+
+    /**
+     * Output a string representation of this sort to a given stream.
+     * @param out the output stream
+     */
+    void toStream(std::ostream& out) const;
+
+    /**
+     * @return a string representation of this sort
+     */
+    std::string toString() const;
+
+  private:
+    /* The interal type wrapped by this sort. */
+    CVC4::Type* d_type;
+};
+
+/**
+ * Serialize a sort to given stream.
+ * @param out the output stream
+ * @param s the sort to be serialized to the given output stream
+ * @return the output stream
+ */
+std::ostream& operator<< (std::ostream& out, const Sort& s) CVC4_PUBLIC;
+
+/**
+ * Hash function for Sorts.
+ */
+struct CVC4_PUBLIC SortHashFunction
+{
+  size_t operator()(const Sort& s) const;
+};
+
+
+/* -------------------------------------------------------------------------- */
+/* Term                                                                       */
+/* -------------------------------------------------------------------------- */
+
+/**
+ * A CVC4 Term.
+ */
+class CVC4_PUBLIC Term
```

Comment:
Have you considered making Term nullable? At least one advantage I can think of is to improve performance of Term() and Term(Term&&). (This is a bit of a pre-mature optimization.)

Reply:
I'm not sure I fully understand what you mean. Term() wraps Expr(), so if Expr::isNull() then Term::isNull(). In Expr we do not handle that in any special way: https://github.com/CVC4/CVC4/blob/master/src/expr/expr_template.cpp#L121
What is it exactly that you are suggesting here?

Reply:
The question was whether `d_expr` could be allowed to be nullptr and to make this equivalent to the current isNull(). This avoids allocs on default allocations, but adds branches. This is a bit of a premature optimization so I would not worry too much about it.